### PR TITLE
Add MKS Instruments 937B vacuum gauge controller.

### PR DESCRIPTION
**Pull request by bilderbuchi**
This adds the instrument from #637 by @dkriegner, with the sensor types properties and the associated needed change to pymeasure core mechanics removed.
I'll replay the mechanics change later, but it's challenging to extract/rebase, I'll probably do it manually (#740).

### CODE REVIEW

**Comment by bilderbuchi** [comment]
@dkriegner, could you please confirm that this works as expected on your end (sans the sensor type stuff)?

File: pymeasure/instruments/mksinst/mks937b.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+    :param adapter: pyvisa resource name of the instrument or adapter instance
+    :param address: device address included in every message to the instrument
+                    (default=253)
+    :param kwargs: Any valid key-word argument for Instrument
+    """
+
+    def __init__(self, adapter, name="MKS 937B vacuum gauge controller", address=253, **kwargs):
+        kwargs.setdefault("write_termination", ";FF")
+        kwargs.setdefault("read_termination", ";")  # in reality its ";FF"
```
Comment (msmttchr) [inline pymeasure/instruments/mksinst/mks937b.py:64]: Do we want to enable overriding of these two parameters?
Just hardcoding them seems ok to me.

File: pymeasure/instruments/mksinst/mks937b.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+    def write(self, command):
+        """
+        Writes to the instrument including the device address
+
+        :param command: command string to be sent to the instrument
+        """
+        super().write(self._prepend_address(command))
+
+    def check_errors(self):
```
Comment (msmttchr) [inline pymeasure/instruments/mksinst/mks937b.py:119]: It's not clear to me when this method is used.

File: pymeasure/instruments/mksinst/mks937b.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                return
+        # no valid acknowledgement message found
+        raise ValueError(f"invalid reply '{ret}' found in check_errors")
+
+    serial = Instrument.measurement(
+        "SN?", """ Serial number of the instrument """,
+        cast=str,
+    )
+
+    pressure1 = Instrument.measurement(
```
Comment (msmttchr) [inline pymeasure/instruments/mksinst/mks937b.py:136]: This and other properties below seem to be channels.
Any reason to not use the new channel mechanism?

**Review by msmttchr — COMMENTED** [review]
Few comments (mostly clarification) from me.

**Review by dkriegner — COMMENTED** [review]
@bilderbuchi should I look into the `Channel` mechanism when testing if this works with the hardware?

File: pymeasure/instruments/mksinst/mks937b.py (hunk 2 above)
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:119]: see properties with `check_set_error` set

File: pymeasure/instruments/mksinst/mks937b.py (hunk 1 above)
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:64]: Yes, if they cannot be changed, hardcode them into `super().__init__`.

**Comment by bmoneke** [comment]
Channels seem a good idea, as the code repeats quite often.
Thanks to the tests, we can see easily, whether the communication stays the same 😁

File: pymeasure/instruments/mksinst/mks937b.py (hunk 2 above)
Comment (msmttchr) [reply pymeasure/instruments/mksinst/mks937b.py:119]: Ok. Thanks for the clarification

**Comment by bilderbuchi** [comment]
>  I assume I have no commit rights here. So the working code can be found at https://github.com/dkriegner/pymeasure/tree/mks937b_no_gauge_type

Thanks for the effort! You can always make a PR against [the branch in my fork](https://github.com/bilderbuchi/pymeasure/tree/PR-637-minimal) that is being used in here -- when I merge it, it will appear in this PR. :-)

**Comment by bilderbuchi** [comment]
Thanks, I merged it. Concerning `Ion gauge status of the channel, only valid for channel 1, 3, and 5` -- I don't know how to best realise this restriction, I think right now if you ask for this on channel 2, you'll get an error from the instrument back? 
Can we redesign this to avoid the problem, or should we implement a different "collection" for the ion gauge properties, @bmoneke @msmttchr ?

**Comment by bmoneke** [comment]
You could make an `ion_gauge` channel (subclass of the other one) and give a list of Channelclasses to the Creator.

**Comment by dkriegner** [comment]
> Thanks, I merged it. Concerning `Ion gauge status of the channel, only valid for channel 1, 3, and 5` -- I don't know how to best realise this restriction, I think right now if you ask for this on channel 2, you'll get an error from the instrument back?

Yes the device answers with a NAK160 -> invalid command.  This is why I included a mapping for this in the ion_gauge_status. I also looked into a way to define different channel types but I thought this means that the prefix would need to be different. and I find it strange to have two different prefixes for the channels just because of this one property. Since no real error is raised if one asks this for channel 2,4,6 I could live with the current solution.

> You could make an `ion_gauge` channel (subclass of the other one) and give a list of Channelclasses to the Creator.

would this work with the same prefix? So could I check the pressure with
```
mks937.ch_1.pressure
mks937.ch_2.pressure
```
?

**Comment by bilderbuchi** [comment]
I implemented that suggestion.

**Comment by bilderbuchi** [comment]
Imo ready for review again.

File: pymeasure/instruments/mksinst/mks937b.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                     "Off": "O",
+                     "Protect": "P",
+                     "Degas": "D",
+                     "Control": "C",
+                     "Rear panel Ctrl off": "R",
+                     "HC filament fault": "H",
+                     "No gauge": "N",
+                     "Good": "G",
+                     "NOT_IONGAUGE": "NAK152",
+                     "INVALID COMMAND": "NAK160",
```
Comment (dkriegner) [inline pymeasure/instruments/mksinst/mks937b.py:41]: The "INVALID COMMAND" line could be dropped now. This should never trigger.

**Review by dkriegner — APPROVED** [review]
I think this one line could still be dropped, but otherwise it looks good to me

**Review by msmttchr — APPROVED** [review]
Ok for me. Thanks

File: pymeasure/instruments/mksinst/mks937b.py (hunk 4 above)
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:41]: I'd keep it in for completeness.

**Review by bmoneke — APPROVED** [review]
Thanks for the work.

Only one small issue.

File: tests/instruments/mksinst/test_mks937b.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def test_pressure():
+    """Verify the communication of the pressure getter."""
+    with expected_protocol(
+        MKS937B,
+        [("@253PR1?", "@253ACK1.10e-9"),
+         (None, b"FF")],
+    ) as inst:
+        inst.adapter.preprocess_reply = inst._extract_reply  # needed to workaround a bug
```
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:37]: good catch. This is indeed not needed anymore. pytest seems to be happy also without this line / these lines.

File: pymeasure/instruments/mksinst/mks937b.py (hunk 4 above)
Comment (bilderbuchi) [reply pymeasure/instruments/mksinst/mks937b.py:41]: Agree, this doesn't hurt to keep.

File: pymeasure/instruments/mksinst/mks937b.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    The communication protocol of this device is as follows:
+
+    Query: '@<aaa><Command>?;FF' with the response '@<aaa>ACK<Response>;FF'
+    Set command: '@<aaa><Command>!<parameter>;FF' with the response '@<aaa>ACK<Response>;FF'
+    Above <aaa> is an address from 001 to 254 which can be specified upon
+    initialization. Since ';FF' is not supported by pyvisa as terminator this
+    class overloads the device communication methods.
+
+    :param adapter: pyvisa resource name of the instrument or adapter instance
+    :param address: device address included in every message to the instrument
```
Comment (dkriegner) [inline pymeasure/instruments/mksinst/mks937b.py:83]:
```suggestion
    :param string name: The name of the instrument.
    :param address: device address included in every message to the instrument
```

File: tests/instruments/mksinst/test_mks937b.py (hunk 5 above)
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:37]:
```suggestion
```

File: tests/instruments/mksinst/test_mks937b.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def test_ion_gauge_status():
+    """Verify the communication of the ion gauge status getter."""
+    with expected_protocol(
+        MKS937B,
+        [("@253T1?", "@253ACKG"),
+         (None, b"FF")],
+    ) as inst:
+        inst.adapter.preprocess_reply = inst._extract_reply  # needed to workaround a bug
```
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:48]:
```suggestion
```

File: tests/instruments/mksinst/test_mks937b.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        assert inst.ch_1.ion_gauge_status == "Good"
+
+
+def test_ion_gauge_status_invalid_channel():
+    """Ion gauge status does not exist on all channels."""
+    with expected_protocol(
+        MKS937B,
+        [],
+    ) as inst:
+        inst.adapter.preprocess_reply = inst._extract_reply  # needed to workaround a bug
```
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:58]:
```suggestion
```

File: tests/instruments/mksinst/test_mks937b.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def test_unit():
+    """Verify the communication of the voltage getter."""
+    with expected_protocol(
+        MKS937B,
+        [("@253U?", "@253ACKTORR"),
+         (None, b"FF")],
+    ) as inst:
+        inst.adapter.preprocess_reply = inst._extract_reply  # needed to workaround a bug
```
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:70]:
```suggestion
```

File: tests/instruments/mksinst/test_mks937b.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def test_power_enabled():
+    """Verify the communication of the voltage getter."""
+    with expected_protocol(
+        MKS937B,
+        [("@253CP1?", "@253ACKON"),
+         (None, b"FF")],
+    ) as inst:
+        inst.adapter.preprocess_reply = inst._extract_reply  # needed to workaround a bug
```
Comment (dkriegner) [inline tests/instruments/mksinst/test_mks937b.py:81]:
```suggestion
```

**Review by dkriegner — CHANGES_REQUESTED** [review]
I hope this makes it easier for you to make the final changes which could be helpful here

**Comment by bilderbuchi** [comment]
> I hope this makes it easier for you to make the final changes which could be helpful here

Yes indeed, thanks for that!

**Comment by bmoneke** [comment]
With that, it is ready to merge, I think. @bilderbuchi

**Comment by bilderbuchi** [comment]
I'm just waiting for the test suite to complete :-)

**Pull request merged**